samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImpl.java

```diff
@@ -29,60 +37,136 @@
+   * @param message  the input message
+   * @param collector  the {@link MessageCollector} for this message
+   * @param coordinator  the {@link TaskCoordinator} for this message
+   */
+  public final void onMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    this.messageCounter.inc();
+    long startNs = System.nanoTime();
+    Collection<RM> results = handleMessage(message, collector, coordinator);
+    long endNs = System.nanoTime();
+    this.handleMessageTimer.update(endNs - startNs);
+
+    results.forEach(rm ->
+        this.registeredOperators.forEach(op ->
+            op.onMessage(rm, collector, coordinator)));
+  }
+
+  /**
+   * Handle the incoming {@code message} and return the results to be propagated to registered operators.
    *
    * @param message  the input message
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
+   * @return  results of the transformation
    */
-  public abstract void onNext(M message, MessageCollector collector, TaskCoordinator coordinator);
+  protected abstract Collection<RM> handleMessage(M message, MessageCollector collector,
+      TaskCoordinator coordinator);
 
   /**
-   * Invoked at every tick. This method delegates to {@link #onTimer(MessageCollector, TaskCoordinator)}
+   * Handle timer ticks for this {@link OperatorImpl} and propagate the results and timer tick to registered operators.
+   * <p>
+   * Delegates to {@link #handleTimer(MessageCollector, TaskCoordinator)} for handling the timer tick.
    *
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
    */
-  public final void onTick(MessageCollector collector, TaskCoordinator coordinator) {
-    onTimer(collector, coordinator);
-    nextOperators.forEach(sub -> sub.onTick(collector, coordinator));
+  public final void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long startNs = System.nanoTime();
+    Collection<RM> results = handleTimer(collector, coordinator);
+    long endNs = System.nanoTime();
+    this.handleTimerTimer.update(endNs - startNs);
+
+    results.forEach(rm ->
+        this.registeredOperators.forEach(op ->
+            op.onMessage(rm, collector, coordinator)));
+    this.registeredOperators.forEach(op -> op.onTimer(collector, coordinator));
   }
 
   /**
-   * Invoked at every tick. Implementations must call {@link #propagateResult} to propagate any generated output
-   * to registered downstream operators.
+   * Handle the the timer tick for this operator and return the results to be propagated to registered operators.
+   * <p>
+   * Defaults to a no-op implementation.
    *
    * @param collector  the {@link MessageCollector} in the context
    * @param coordinator  the {@link TaskCoordinator} in the context
+   * @return  results of the timed operation
```

Comment:
Ah, this is my moment of zen. I did not know that `Collections.emptyList` returns a cached, immutable list.